pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -454,6 +454,22 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun, get
 			pipelineMeta.Namespace, pipelineMeta.Name, err)
 		return controller.NewPermanentError(err)
 	}
+	// If the pipeline spec was fetched from a referenced pipeline then we also need to validate
+	// the parameter and workspace usage because propagation is not allowed.
+	if pr.Spec.PipelineRef != nil {
+		if err := pipelineSpec.ValidatePipelineParameterUsage(ctx); err != nil {
+			pr.Status.MarkFailed(ReasonFailedValidation,
+				"Pipeline %s/%s can't be Run; it has an invalid param usage: %s",
+				pipelineMeta.Namespace, pipelineMeta.Name, err)
+			return controller.NewPermanentError(err)
+		}
+		if err := pipelineSpec.ValidatePipelineWorkspacesUsage(); err != nil {
```

Comment:
Just curious if we could put those 2 validation functions ValidatePipelineParameterUsage and ValidatePipelineWorkspacesUsage inside pipelineSpec.Validate()? Since we call pipelineSpec.Validate above so we don't need to call them again.

Reply:
We also call these in the reconciler so we can't merge it in I think.

Reply:
We cannot do that since params can be propagated in the normal case so it's possible that the parameters are not declared explicitly in the embedded pipeline Spec. Here, when we use remote resolution, we expect the declaration to be explicit, hence we do the check here separately.